pisa/stages/discr_sys/csv_hypersurfaces.py

```diff
@@ -0,0 +1,277 @@
+import ast
+from collections.abc import Mapping
+import numpy as np
+import pandas as pd
+
+from pisa import FTYPE, ureg
+from pisa.core.binning import OneDimBinning, MultiDimBinning
+from pisa.core.param import Param, ParamSet
+from pisa.core.stage import Stage
+from pisa.utils.log import logging
+from pisa.utils.format import split
+from pisa.utils.resources import find_resource
+
+__all__ = ["csv_hypersurfaces",]
+
+__author__ = "B. Benkel, J. Weldert"
+
+__license__ = """Copyright (c) 2014-2025, The IceCube Collaboration
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License."""
+
+
+class csv_hypersurfaces(Stage):
+    """
+    Service to apply hypersurface parameterisation produced by
+    `scripts.fit_discrete_sys_nd`
+
+    Parameters
+    ----------
+    fit_results_file : str
+        Path to hypersurface fit results file(s)
+
+    links : dict
+        A dictionary defining how containers should be linked. Keys are the names of
+        the merged containers, values are lists of containers being linked together.
+        Keys must be a sub-set of the loaded hypersurfaces.
+
+    nominal_systematics : dict
+        Systematics and their nominal values
+        
+    inter_param : str
+        Parameter used for interpolation
+
+    propagate_uncertainty : bool, optional
+        Propagate the uncertainties from the hypersurface to the uncertainty of
+        the output
+    """
+    def __init__(
+        self,
+        fit_results_file,
+        nominal_systematics,
+        inter_param,
+        links=None,
+        propagate_uncertainty=False,
+        **std_kwargs,
+    ):
+        # -- Only allowed/implemented modes -- #
+        assert isinstance(std_kwargs['calc_mode'], MultiDimBinning)
+        # -- Load hypersurfaces -- #
+
+        # Store args
+        self.fit_results_file = split(fit_results_file)
+        
+        if isinstance(nominal_systematics, str):
+            self.nominal_systematics = eval(nominal_systematics)
+        elif isinstance(nominal_systematics, dict):
+            self.nominal_systematics = nominal_systematics
+        else:
+            raise ValueError(
+                f"Unsupported type {type(nominal_systematics)} for nominal_systematics."
+            )
+        
+        self.inter_param = inter_param
+        self.propagate_uncertainty = propagate_uncertainty
+
+        expected_container_keys = [
+            'weights',
+        ]
+        if 'error_method' in std_kwargs:
+            if std_kwargs['error_method']:
```

Comment:
can be moved into line above: if `error_method` in std_kwargs and `std_kwargs['error_method']:` since second condition only ever checked if first is satisfied